chpl/chpl-service/src/main/java/gov/healthit/chpl/listing/measure/ListingMeasureDAO.java

```diff
@@ -0,0 +1,226 @@
+
+        ListingMeasureEntity mmEntity = new ListingMeasureEntity();
+        mmEntity.setDeleted(false);
+        mmEntity.setLastModifiedUser(AuthUtil.getAuditId());
+        mmEntity.setListingId(listingId);
+        mmEntity.setMeasureId(mm.getMeasure().getId());
+        mmEntity.setTypeId(mm.getMeasurementType().getId());
+        create(mmEntity);
+
+        for (CertificationCriterion associatedCriterion : mm.getAssociatedCriteria()) {
+            ListingMeasureCriterionMapEntity mmCriterionEntity = new ListingMeasureCriterionMapEntity();
+            mmCriterionEntity.setCertificationCriterionId(associatedCriterion.getId());
+            mmCriterionEntity.setListingMeasureMapId(mmEntity.getId());
+            mmCriterionEntity.setDeleted(false);
+            mmCriterionEntity.setLastModifiedUser(AuthUtil.getAuditId());
+            create(mmCriterionEntity);
+        }
+    }
+
+    public void updateCertifiedProductMeasureMapping(ListingMeasure toUpdate)
+            throws EntityRetrievalException {
+
+        ListingMeasureEntity existingEntity = getEntityById(toUpdate.getId());
+        if (existingEntity == null) {
+            throw new EntityRetrievalException("Could not find mapping with id " + toUpdate.getId());
+        }
+        existingEntity.setMeasureId(toUpdate.getMeasure().getId());
+        existingEntity.setTypeId(toUpdate.getMeasurementType().getId());
+        existingEntity.setLastModifiedUser(AuthUtil.getAuditId());
+        update(existingEntity);
+
+        Optional<List<CertificationCriterion>> updatedAssociatedCriteria = Optional
+                .ofNullable(new ArrayList(toUpdate.getAssociatedCriteria()));
+        Optional<List<CertificationCriterion>> exsitingAssociatedCriteria = Optional
+                .ofNullable(existingEntity.getAssociatedCriteria().stream()
+                        .map(assocCriterion -> assocCriterion.convert())
+                        .collect(Collectors.toList()));
+
+        List<CertificationCriterion> removed = getRemovedAssociatedCriteria(updatedAssociatedCriteria,
+                exsitingAssociatedCriteria);
+
+        List<CertificationCriterion> added = getAddedAssociatedCriteria(updatedAssociatedCriteria,
+                exsitingAssociatedCriteria);
+
+        removed.stream().forEach(removedCriterion -> {
+            Optional<ListingMeasureCriterionMapEntity> removedEntityOpt =
+                    existingEntity.getAssociatedCriteria().stream()
+                        .filter(existingCrit -> existingCrit.getCriterion().getId().equals(removedCriterion.getId()))
+                        .findAny();
+            if (removedEntityOpt.isPresent()) {
+                ListingMeasureCriterionMapEntity removedEntity = removedEntityOpt.get();
+                removedEntity.setDeleted(true);
+                removedEntity.setLastModifiedUser(AuthUtil.getAuditId());
+                update(removedEntity);
+            }
+        });
```

Comment:
For readability, you might want to pull the code in the `foreach` to a new method with that is named based on what is happening.  It could really improve the readability.
```
removed.stream()
        .foreach(removedCriteria -> deleteCriteria(removedCriteria));
```

Reply:
Fixed